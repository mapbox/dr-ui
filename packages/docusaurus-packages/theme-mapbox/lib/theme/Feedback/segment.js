import forwardEvent from './forward-event';
import * as Sentry from '@sentry/browser';
import env from '@theme/AnalyticsShell/env';

// Send the event to Segment with forward event
export function sendToSegment({ state, props }) {
  const event = createSegmentEvent({ state, props });
  forwardEvent(event, props.webhook, (err) => {
    if (err) Sentry.captureException(err);
  });
}

// Creates event to send to Segment
function createSegmentEvent({ state, props }) {
  const { user, anonymousId, feedback, sessionId, category, categoryType } =
    state;
  const { site, section, location } = props;
  // set user if available else set anonymousId (needed for forward-event request)
  const identity = {
    ...(user && user.id ? { userId: user.id } : { anonymousId: anonymousId })
  };

  return {
    event: 'Sent docs feedback',
    ...identity,
    properties: {
      // the unique id for the feedbak session
      sessionId,
      // the feedback category that the user selected (helpful for yes, unhelpful for no)
      category,
      // the sub type for the category
      categoryType,
      // text feedback, if available
      ...(feedback && { feedback: feedback }),
      // name of current site (important for filtering in Mode)
      site,
      // (optional) name of section for longer pagers, helpful for fitering in Mode and identifying section areas
      section,
      // get page context
      page: location || undefined,
      // set user if available else set anonymousId (needed for Mode)
      userId: user && user.id ? user.id : undefined,
      anonymousId: anonymousId,
      // set plan, if available
      ...(user && user.plan && { plan: user.plan.id }),
      // get environment: staging or production
      environment: env()
    }
  };
}
