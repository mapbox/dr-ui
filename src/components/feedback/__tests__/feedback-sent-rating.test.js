import React from 'react';
import Feedback from '..';
import toJson from 'enzyme-to-json';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('Sent helpful rating - yes', () => {
  const feedback = mount(
    <Feedback
      site="dr-ui"
      location={{
        pathname: '/mapbox-gl-js/api/',
        hash: '#lnglat'
      }}
      userName="decorah"
      webhook={{
        production: '',
        staging: ''
      }}
    />
  );
  test('clicked yes', () => {
    feedback.find('#dr-ui--feedback-page-yes').simulate('click');
    expect(feedback.state()).toEqual({
      event: {
        event: 'Sent docs feedback',
        userId: 'decorah',
        properties: {
          environment: 'staging',
          helpful: true,
          location: {
            hash: '',
            host: 'localhost',
            hostname: 'localhost',
            href: 'http://localhost/',
            origin: 'http://localhost',
            pathname: '/',
            search: ''
          },
          page: { hash: '#lnglat', pathname: '/mapbox-gl-js/api/' },
          section: undefined,
          site: 'dr-ui',
          userId: 'decorah'
        }
      },
      feedback: undefined,
      feedbackSent: undefined,
      helpful: true
    });
  });
});

describe('Sent helpful rating - no', () => {
  const feedback = mount(
    <Feedback
      site="dr-ui"
      location={{
        pathname: '/mapbox-gl-js/api/',
        hash: '#lnglat'
      }}
      userName="decorah"
      webhook={{
        production: '',
        staging: ''
      }}
    />
  );
  test('clicked no', () => {
    feedback.find('#dr-ui--feedback-page-no').simulate('click');
    expect(feedback.state()).toEqual({
      event: {
        event: 'Sent docs feedback',
        userId: 'decorah',
        properties: {
          environment: 'staging',
          helpful: false,
          location: {
            hash: '',
            host: 'localhost',
            hostname: 'localhost',
            href: 'http://localhost/',
            origin: 'http://localhost',
            pathname: '/',
            search: ''
          },
          page: { hash: '#lnglat', pathname: '/mapbox-gl-js/api/' },
          section: undefined,
          site: 'dr-ui',
          userId: 'decorah'
        }
      },
      feedback: undefined,
      feedbackSent: undefined,
      helpful: false
    });
  });
});

describe('Sent helpful rating snapshot', () => {
  const feedback = shallow(
    <Feedback
      site="dr-ui"
      location={{
        pathname: '/mapbox-gl-js/api/',
        hash: '#lnglat'
      }}
    />
  );
  feedback.setState({
    helpful: true
  });

  test('snapshot', () => {
    expect(toJson(feedback)).toMatchSnapshot();
  });
});
