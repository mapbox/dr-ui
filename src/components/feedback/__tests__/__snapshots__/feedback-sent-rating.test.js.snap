// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback character limit Character limit ok, user can submit feedback 1`] = `
<Feedback
  feedbackSentryDsn="https://eccc8b561b9a461990309b01d33d54e3@sentry.io/1848287"
  location={Object {}}
  site="dr-ui"
  type="page"
>
  <div
    className="dr-ui--feedback"
  >
    <AsideHeading>
      <h2
        className="unprose txt-m txt-bold mb6"
      >
        Thanks for your feedback.
      </h2>
    </AsideHeading>
    <div
      className="mt12"
    >
      <div
        className="mb6"
      >
        If you have more specific feedback
        , you can provide it below (optional):
      </div>
      <div
        className="relative"
      >
        <ControlTextarea
          id="dr-ui--feedback-page-text"
          noAuto={false}
          onChange={[Function]}
          optional={false}
          themeControlTextarea="textarea hmin120"
          themeControlWrapper="bg-white mb6"
          value="Lorem ipsum dolor sit amet"
        >
          <ControlWrapper
            id="dr-ui--feedback-page-text"
            themeControlWrapper="bg-white mb6"
          >
            <div
              className="relative bg-white mb6"
            >
              <textarea
                aria-required={true}
                className="textarea hmin120"
                data-test="dr-ui--feedback-page-text-textarea"
                id="dr-ui--feedback-page-text"
                name="dr-ui--feedback-page-text"
                onChange={[Function]}
                value="Lorem ipsum dolor sit amet"
              />
              <div
                role="alert"
              />
            </div>
          </ControlWrapper>
        </ControlTextarea>
        <FeedbackCounter
          createId={[Function]}
          feedbackLength={974}
          feedbackOverLimit={false}
        >
          <div
            className="absolute bottom right mb6 mr18 txt-mono bg-lighten75 px3 txt-s"
            id="dr-ui--feedback-page-counter"
          >
            974
          </div>
        </FeedbackCounter>
      </div>
      <div
        className="mb12"
      >
        <button
          className="btn btn--s mb6 mr12 inline-block"
          disabled={false}
          id="dr-ui--feedback-page-submit"
          onClick={[Function]}
        >
          Send feedback
        </button>
      </div>
      <div
        className="txt-s txt-em"
      >
        This form is for documentation feedback. If you have a technical question about how to use a Mapbox product,
         
        <a
          className="link"
          href="https://support.mapbox.com/hc/en-us"
        >
          contact Support
        </a>
        .
      </div>
    </div>
  </div>
</Feedback>
`;

exports[`Feedback character limit Character limit too long, user cannot submit feedback 1`] = `
<Feedback
  feedbackSentryDsn="https://eccc8b561b9a461990309b01d33d54e3@sentry.io/1848287"
  location={Object {}}
  site="dr-ui"
  type="page"
>
  <div
    className="dr-ui--feedback"
  >
    <AsideHeading>
      <h2
        className="unprose txt-m txt-bold mb6"
      >
        Thanks for your feedback.
      </h2>
    </AsideHeading>
    <div
      className="mt12"
    >
      <div
        className="mb6"
      >
        If you have more specific feedback
        , you can provide it below (optional):
      </div>
      <div
        className="relative"
      >
        <ControlTextarea
          id="dr-ui--feedback-page-text"
          noAuto={false}
          onChange={[Function]}
          optional={false}
          themeControlTextarea="textarea hmin120"
          themeControlWrapper="bg-white mb6"
          value="Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."
        >
          <ControlWrapper
            id="dr-ui--feedback-page-text"
            themeControlWrapper="bg-white mb6"
          >
            <div
              className="relative bg-white mb6"
            >
              <textarea
                aria-required={true}
                className="textarea hmin120"
                data-test="dr-ui--feedback-page-text-textarea"
                id="dr-ui--feedback-page-text"
                name="dr-ui--feedback-page-text"
                onChange={[Function]}
                value="Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."
              />
              <div
                role="alert"
              />
            </div>
          </ControlWrapper>
        </ControlTextarea>
        <FeedbackCounter
          createId={[Function]}
          feedbackLength={-340}
          feedbackOverLimit={true}
        >
          <div
            className="absolute bottom right mb6 mr18 txt-mono bg-lighten75 px3 txt-s color-red"
            id="dr-ui--feedback-page-counter"
          >
            -340
          </div>
        </FeedbackCounter>
      </div>
      <div
        className="mb12"
      >
        <button
          className="btn btn--s mb6 mr12 inline-block"
          disabled={true}
          id="dr-ui--feedback-page-submit"
          onClick={[Function]}
        >
          Send feedback
        </button>
        <FeedbackOverlimitWarning
          createId={[Function]}
          feedbackLimit={1000}
        >
          <span
            className="color-red txt-s bg-red-faint round inline-block py3 px12"
            id="dr-ui--feedback-page-overlimit"
          >
            <Icon
              inline={true}
              name="alert"
              passthroughProps={Object {}}
              size={18}
            >
              <svg
                className="events-none icon inline-block align-t"
                focusable="false"
                role="presentation"
                style={
                  Object {
                    "height": 18,
                    "width": 18,
                  }
                }
              >
                <use
                  xlinkHref="#icon-alert"
                  xmlnsXlink="http://www.w3.org/1999/xlink"
                />
              </svg>
            </Icon>
             Your message is over the
             
            1000
             character limit.
          </span>
        </FeedbackOverlimitWarning>
      </div>
      <div
        className="txt-s txt-em"
      >
        This form is for documentation feedback. If you have a technical question about how to use a Mapbox product,
         
        <a
          className="link"
          href="https://support.mapbox.com/hc/en-us"
        >
          contact Support
        </a>
        .
      </div>
    </div>
  </div>
</Feedback>
`;

exports[`Sent helpful rating snapshot snapshot 1`] = `
<div
  className="dr-ui--feedback"
>
  <AsideHeading>
    Thanks for your feedback.
  </AsideHeading>
  <div
    className="mt12"
  >
    <div
      className="mb6"
    >
      If you have more specific feedback
      , you can provide it below (optional):
    </div>
    <div
      className="relative"
    >
      <ControlTextarea
        id="dr-ui--feedback-page-text"
        noAuto={false}
        onChange={[Function]}
        optional={false}
        themeControlTextarea="textarea hmin120"
        themeControlWrapper="bg-white mb6"
        value=""
      />
      <FeedbackCounter
        createId={[Function]}
        feedbackLength={1000}
        feedbackOverLimit={false}
      />
    </div>
    <div
      className="mb12"
    >
      <button
        className="btn btn--s mb6 mr12 inline-block"
        disabled={true}
        id="dr-ui--feedback-page-submit"
        onClick={[Function]}
      >
        Send feedback
      </button>
    </div>
    <div
      className="txt-s txt-em"
    >
      This form is for documentation feedback. If you have a technical question about how to use a Mapbox product,
       
      <a
        className="link"
        href="https://support.mapbox.com/hc/en-us"
      >
        contact Support
      </a>
      .
    </div>
  </div>
</div>
`;
